python/ray/_private/client_mode_hook.py

```diff
@@ -1,29 +1,40 @@
 import os
 from contextlib import contextmanager
 from functools import wraps
+import threading
 
 # Attr set on func defs to mark they have been converted to client mode.
 RAY_CLIENT_MODE_ATTR = "__ray_client_mode_key__"
 
 client_mode_enabled = os.environ.get("RAY_CLIENT_MODE", "0") == "1"
 os.environ.update({"RAY_CLIENT_MODE": "0"})
 
-_client_hook_enabled = True
+_client_hook_enabled = threading.local()
+_client_hook_enabled.val = True
+
+def _get_client_hook_enabled():
+    global _client_hook_enabled
+    if not hasattr(_client_hook_enabled, "val"):
+        _client_hook_enabled.val = True
+    return _client_hook_enabled.val
 
 
 def _enable_client_hook(val: bool):
     global _client_hook_enabled
-    _client_hook_enabled = val
+    _client_hook_enabled.val = val
 
 
 def _disable_client_hook():
     global _client_hook_enabled
-    out = _client_hook_enabled
-    _client_hook_enabled = False
+    out = _get_client_hook_enabled()
+    _client_hook_enabled.val = False
     return out
 
 
 def _explicitly_enable_client_mode():
+    """Force client mode to be enabled.
+    NOTE: This should not be used in tests, use `enable_client_mode`.
+    """
     global client_mode_enabled
```

Comment:
What is the difference between client mode enabled and client hook enabled?

Reply:
I think "client mode enabled" is  a global setting saying whether ray client is enabled and "client hook enabled" is an internal toggle used by the code to turn on/off delegating ray apis to the client stub.

It could help to add a comment explaining the meanings to this file.

Reply:
Yep as @DmitriGekhtman said. 

I added two comments explaining these!!

I also renamed them:
* `client_mode_enabled` -> `is_client_mode_enabled` (making it clear that this is answering the question of 'is client mode currently enabled'.
* `client_hook_enabled` -> `_client_hook_status_on_thread` (making it clear that it is thread specific).